File: precipitation/precipitation_nicolosi.R
# SDSU Climate Informatics Lab
# San Diego State University
# by Waverley Moody
# Supervised by Distinguished Professor Samuel Shen
# R Code Version 1.0.0
#
# A reproduction of the University of Washington General Circulation
# Animations Library by Professor John Michael Wallace.
#
# Script: precipitation_nicolosi.R
# Description: Generates the precipitation climatology animation from ERA5 reanalysis data (1979-2000), rendered as a double-hemisphere Nicolosi Globular projection. Weekly totals (cm water equivalent) computed server-side by the derived-era5-single-levels-daily-statistics CDS dataset, averaged into a 22-year climatological mean per 48-frame weekly cycle.
#
# Note: For the Plate Carrée, Robinson, and Foucaut projections, see the other scripts in the precipitation scripts folder.

library(terra)
library(ncdf4)
library(rnaturalearth)
library(sf)
library(av)

# Disable S2 spherical geometry for hemisphere clipping.
sf::sf_use_s2(FALSE)

# ── Paths ────────────────────────────────────────────────────────────────

data_file <- "/Volumes/CLIMATEDATA/precip/climatology/climatology_precip_48frame.nc"

output_dir <- paste0(
  "/Users/waverleymoody/Downloads/",
  "precipitation_animation/r_output"
)

dir.create(
  output_dir,
  showWarnings = FALSE,
  recursive = TRUE
)

frames_dir <- file.path(
  output_dir,
  "nicolosi",
  "frames"
)

dir.create(
  frames_dir,
  showWarnings = FALSE,
  recursive = TRUE
)

# ── Projection settings ──────────────────────────────────────────────────

WEST_LON <- -90
EAST_LON <- 90

# Higher values create smoother output but increase processing time.
grid_res <- 600

# ── Plot settings ────────────────────────────────────────────────────────

vmin <- 0
vmax <- 21
tick_step <- 3

# Same gradient as the Python Plate Carrée/Robinson/Foucaut/Nicolosi
# precipitation scripts: white at 0 cm blends into blue over just the
# first 0.3 cm, then continues through blue -> green -> yellow -> orange
# -> red -> purple up to 21 cm. Positions are fractions of vmax, mirroring
# the Python color_stops list exactly.
precip_colors <- c(
  "#ffffff", # white           (0 cm)
  "#d0e8f0", # pale blue       (0.3 cm)
  "#7ab8e0", # light blue      (2 cm)
  "#3060c0", # blue            (3 cm)
  "#00994d", # green           (6 cm)
  "#66cc00", # yellow-green    (8 cm)
  "#ffff00", # yellow          (9 cm)
  "#ff9900", # orange          (10 cm)
  "#ff3300", # red-orange      (11 cm)
  "#990000", # dark red        (12 cm)
  "#660066"  # purple          (21 cm)
)
precip_positions <- c(0, 0.3, 2, 3, 6, 8, 9, 10, 11, 12, 21) / vmax

# Mirrors LinearSegmentedColormap.from_list with (position, color) tuples,
# since colorRampPalette() only supports evenly-spaced stops -- needed
# here (unlike the temperature Nicolosi script) because precipitation's
# color stops are deliberately uneven (white compressed into just the
# first 0.3 cm out of a 21 cm range).
make_custom_palette <- function(colors, positions, n) {
  rgb_mat  <- t(col2rgb(colors)) / 255
  eval_pos <- seq(0, 1, length.out = n)
  r <- approx(positions, rgb_mat[, 1], xout = eval_pos)$y
  g <- approx(positions, rgb_mat[, 2], xout = eval_pos)$y
  b <- approx(positions, rgb_mat[, 3], xout = eval_pos)$y
  rgb(r, g, b)
}

n_colors <- 60  # matches the Python script's 60 color steps

precip_palette <- make_custom_palette(
  precip_colors,
  precip_positions,
  n_colors
)

breaks <- seq(
  vmin,
  vmax,
  length.out = n_colors + 1
)

# ── Natural Earth basemap layers ─────────────────────────────────────────

cat("Downloading/loading basemap layers...\n")

coastline <- ne_coastline(
  scale = 50,
  returnclass = "sf"
)

borders <- ne_countries(
  scale = 50,
  returnclass = "sf"
)

# ── Load pre-computed climatology ────────────────────────────────────────

r_stack <- rast(
  data_file,
  subds = "tp_weekly_total_cm"
)

# Convert ERA5 longitudes from 0-360 to -180-180.
r_stack <- rotate(r_stack)

# Already in cm -- no unit conversion needed (unlike temperature's
# Kelvin -> Celsius step here).

n_frames <- nlyr(r_stack)

lons <- xFromCol(
  r_stack,
  1:ncol(r_stack)
)

lats <- yFromRow(
  r_stack,
  1:nrow(r_stack)
)

nc <- nc_open(
  data_file
)

frame_labels <- ncvar_get(
  nc,
  "frame_label"
)

nc_close(nc)

cat(
  sprintf(
    "Loaded %d frames, grid %d x %d\n",
    n_frames,
    length(lats),
    length(lons)
  )
)

# ── Build one hemisphere ─────────────────────────────────────────────────

build_hemisphere <- function(
    lons,
    lats,
    central_lon,
    grid_res
) {
  
  proj_str <- sprintf(
    "+proj=nicol +lon_0=%f +R=6371000",
    central_lon
  )
  
  # Calculate angular longitude offset from the hemisphere center.
  lon_diff <- (
    (lons - central_lon + 180) %% 360
  ) - 180
  
  # Keep only longitudes within 90 degrees of the center.
  mask_1d <- abs(lon_diff) <= 90
  
  candidate_idx <- which(
    mask_1d
  )
  
  # Sort longitudes by angular offset.
  ord <- order(
    lon_diff[candidate_idx]
  )
  
  col_index <- candidate_idx[ord]
  
  lons_hemi <- lons[col_index]
  
  # Create the longitude/latitude grid.
  #
  # Latitude varies fastest, matching R's column-major matrix flattening
  # when as.vector() is used later.
  grid <- expand.grid(
    lat = lats,
    lon = lons_hemi
  )
  
  pts <- vect(
    cbind(
      grid$lon,
      grid$lat
    ),
    crs = "EPSG:4326"
  )
  
  XY <- crds(
    project(
      pts,
      proj_str
    )
  )
  
  # Trace the outer boundary of the hemisphere.
  edge_lats <- seq(
    -90,
    90,
    length.out = 400
  )
  
  b1 <- vect(
    cbind(
      rep(
        central_lon + 90,
        length(edge_lats)
      ),
      edge_lats
    ),
    crs = "EPSG:4326"
  )
  
  b2 <- vect(
    cbind(
      rep(
        central_lon - 90,
        length(edge_lats)
      ),
      edge_lats
    ),
    crs = "EPSG:4326"
  )
  
  b1_xy <- crds(
    project(
      b1,
      proj_str
    )
  )
  
  b2_xy <- crds(
    project(
      b2,
      proj_str
    )
  )
  
  boundary_xy <- rbind(
    b1_xy,
    b2_xy[nrow(b2_xy):1, ]
  )
  
  boundary_poly <- vect(
    rbind(
      boundary_xy,
      boundary_xy[1, ]
    ),
    type = "polygons",
    crs = proj_str
  )
  
  # Clip coastlines and borders before projection.
  hemi_bbox <- st_as_sfc(
    st_bbox(
      c(
        xmin = central_lon - 90,
        xmax = central_lon + 90,
        ymin = -90,
        ymax = 90
      ),
      crs = st_crs(4326)
    )
  )
  
  coastline_clip <- suppressWarnings(
    st_intersection(
      st_geometry(coastline),
      hemi_bbox
    )
  )
  
  borders_clip <- suppressWarnings(
    st_intersection(
      st_geometry(borders),
      hemi_bbox
    )
  )
  
  coastline_p <- st_transform(
    coastline_clip,
    proj_str
  )
  
  borders_p <- st_transform(
    borders_clip,
    proj_str
  )
  
  # Create the regular projected-space raster template.
  template <- rast(
    ext(boundary_poly),
    ncol = grid_res,
    nrow = grid_res,
    crs = proj_str
  )
  
  list(
    col_index = col_index,
    XY = XY,
    proj_str = proj_str,
    template = template,
    boundary_poly = boundary_poly,
    coastline_p = coastline_p,
    borders_p = borders_p
  )
}

cat("Precomputing Western Hemisphere geometry...\n")

west <- build_hemisphere(
  lons = lons,
  lats = lats,
  central_lon = WEST_LON,
  grid_res = grid_res
)

cat("Precomputing Eastern Hemisphere geometry...\n")

east <- build_hemisphere(
  lons = lons,
  lats = lats,
  central_lon = EAST_LON,
  grid_res = grid_res
)

# ── Rasterize one hemisphere for one animation frame ─────────────────────

render_hemisphere_raster <- function(
    frame_mat,
    hemi
) {
  
  # Extract the data columns belonging to this hemisphere.
  data_hemi <- frame_mat[
    ,
    hemi$col_index,
    drop = FALSE
  ]
  
  # Flatten the matrix in column-major order.
  values_vec <- as.vector(
    data_hemi
  )
  
  valid <- (
    is.finite(values_vec) &
      is.finite(hemi$XY[, 1]) &
      is.finite(hemi$XY[, 2])
  )
  
  pts_vals <- vect(
    hemi$XY[
      valid,
      ,
      drop = FALSE
    ],
    crs = hemi$proj_str
  )
  
  pts_vals$val <- values_vec[valid]
  
  # Rasterize the projected point values.
  r_hemi <- rasterize(
    pts_vals,
    hemi$template,
    field = "val",
    fun = "mean"
  )
  
  # Fill empty cells caused by uneven projected point spacing.
  filled <- focal(
    r_hemi,
    w = 5,
    fun = "mean",
    na.rm = TRUE,
    na.policy = "only"
  )
  
  r_hemi <- cover(
    r_hemi,
    filled
  )
  
  # Clip the result to the hemisphere boundary.
  mask(
    r_hemi,
    hemi$boundary_poly
  )
}

# ── Generate PNG frames ──────────────────────────────────────────────────

frame_paths <- character(
  n_frames
)

for (i in seq_len(n_frames)) {
  
  frame_mat <- as.matrix(
    r_stack[[i]],
    wide = TRUE
  )
  
  west_r <- render_hemisphere_raster(
    frame_mat,
    west
  )
  
  east_r <- render_hemisphere_raster(
    frame_mat,
    east
  )
  
  label <- frame_labels[i]
  
  fpath <- file.path(
    frames_dir,
    sprintf(
      "frame_%03d.png",
      i - 1
    )
  )
  
  png(
    filename = fpath,
    width = 1800,
    height = 980,
    res = 150
  )
  
  # Layout:
  #
  #   Panel 1: full-width title row
  #   Panel 2: western hemisphere
  #   Panel 3: eastern hemisphere
  #   Panel 4: full-width colorbar
  #
  # The first value controls title-row height.
  # The third value controls colorbar-row height.
  layout(
    matrix(
      c(
        1, 1,
        2, 3,
        4, 4
      ),
      nrow = 3,
      byrow = TRUE
    ),
    heights = c(
      0.42,
      5,
      0.95
    )
  )
  
  # ── Full-width title panel ─────────────────────────────────────────────
  
  par(
    mar = c(0, 1, 0, 1),
    bty = "n"
  )
  
  plot.new()
  
  # Left-side title.
  mtext(
    "ERA-5 | Climate Reanalyzer",
    side = 3,
    line = -1.8,
    adj = 0,
    cex = 1.0,
    font = 2
  )
  
  # Right-side title.
  mtext(
    sprintf(
      "%s; 1979\u20132000 Weekly Total Mean",
      label
    ),
    side = 3,
    line = -1.8,
    adj = 1,
    cex = 1.0,
    font = 2
  )
  
  # ── Western Hemisphere panel ───────────────────────────────────────────
  
  par(
    mar = c(1, 1, 1, 1),
    bty = "n"
  )
  
  plot(
    west_r,
    col = precip_palette,
    breaks = breaks,
    range = c(vmin, vmax),
    axes = FALSE,
    legend = FALSE,
    main = "",
    asp = 1
  )
  
  plot(
    west$coastline_p,
    add = TRUE,
    lwd = 1.0,
    col = "black"
  )
  
  plot(
    west$borders_p,
    add = TRUE,
    lwd = 0.8,
    border = "black"
  )
  
  plot(
    west$boundary_poly,
    add = TRUE,
    border = "black",
    lwd = 1,
    col = NA
  )
  
  # ── Eastern Hemisphere panel ───────────────────────────────────────────
  
  par(
    mar = c(1, 1, 1, 1),
    bty = "n"
  )
  
  plot(
    east_r,
    col = precip_palette,
    breaks = breaks,
    range = c(vmin, vmax),
    axes = FALSE,
    legend = FALSE,
    main = "",
    asp = 1
  )
  
  plot(
    east$coastline_p,
    add = TRUE,
    lwd = 1.0,
    col = "black"
  )
  
  plot(
    east$borders_p,
    add = TRUE,
    lwd = 0.8,
    border = "black"
  )
  
  plot(
    east$boundary_poly,
    add = TRUE,
    border = "black",
    lwd = 1,
    col = NA
  )
  
  # ── Full-width colorbar panel ──────────────────────────────────────────
  
  par(
    mar = c(4.0, 6, 0.8, 6),
    bty = "n",
    pty = "m"
  )
  
  pad <- (
    vmax - vmin
  ) * 0.25
  
  cbar_z <- matrix(
    seq_len(n_colors),
    ncol = 1
  )
  
  # x = breaks (true cell edges, length n_colors+1) rather than n_colors
  # midpoints -- this is the edge-vs-center fix from the Foucaut script:
  # passing midpoints makes image() auto-extend the painted area half a
  # bin-width past both vmin and vmax, letting the purple end bleed past
  # the border rect() drawn at the true vmax.
  image(
    x = breaks,
    y = c(0.7, 1.3),
    z = cbar_z,
    col = precip_palette,
    breaks = seq(
      0.5,
      n_colors + 0.5,
      by = 1
    ),
    axes = FALSE,
    xlab = "",
    ylab = "",
    xlim = c(
      vmin - pad,
      vmax + pad
    ),
    ylim = c(
      0.5,
      1.5
    )
  )
  
  rect(
    vmin,
    0.7,
    vmax,
    1.3,
    border = "black",
    lwd = 1.2
  )
  
  axis(
    side = 1,
    at = seq(
      vmin,
      vmax,
      tick_step
    ),
    labels = seq(
      vmin,
      vmax,
      tick_step
    ),
    cex.axis = 1.0
  )
  
  mtext(
    "Precipitation (cm)",
    side = 1,
    line = 2.7,
    cex = 1.0
  )
  
  # Close the PNG device.
  dev.off()
  
  # Save the completed frame path.
  frame_paths[i] <- fpath
  
  cat(
    sprintf(
      "Saved frame %d/%d\n",
      i,
      n_frames
    )
  )
}

# ── Encode frames as MP4 ─────────────────────────────────────────────────

video_path <- file.path(
  output_dir,
  "precipitation_climatology_nicolosi.mp4"
)

av_encode_video(
  input = frame_paths,
  output = video_path,
  framerate = 4,
  codec = "libx264"
)

cat(
  sprintf(
    "Done! Animation saved to: %s\n",
    video_path
  )
)

File: precipitation/precipitation_nicolosi.py
"""
SDSU Climate Informatics Lab
San Diego State University
by Waverley Moody
Supervised by Distinguished Professor Samuel Shen
Python Code Version 1.0.0

A reproduction of the University of Washington General Circulation
Animations Library by Professor John Michael Wallace.

Script: precipitation_nicolosi.py
Description: Generates the precipitation climatology animation from ERA5 reanalysis
data (1979-2000), rendered as a double-hemisphere Nicolosi Globular projection. 
Weekly totals (cm water equivalent) computed server-side by the derived-era5-single-levels-daily-statistics 
CDS dataset, averaged into a 22-year climatological mean per 48-frame weekly cycle.
Note: For the Plate Carrée, Robinson, and Foucaut projections, see the other scripts in the precipitation scripts folder.
"""

import numpy as np
import xarray as xr
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
from pyproj import Transformer
from shapely.geometry import box
from shapely.ops import transform as shp_transform
import cartopy.io.shapereader as shpreader
import imageio.v2 as imageio
import imageio_ffmpeg
import io
from pathlib import Path

matplotlib.rcParams['font.family'] = 'Arial'

# ---- config ------------------------------------------------------------
NC_PATH = "/Volumes/CLIMATEDATA/precip/climatology/climatology_precip_48frame.nc"
OUTPUT_DIR = Path("/Users/waverleymoody/Downloads/precipitation_animation")
OUTPUT_DIR.mkdir(exist_ok=True, parents=True)
OUT_MP4_PATH = str(OUTPUT_DIR / "precipitation_climatology_nicolosi.mp4")
WEST_LON = -90.0
EAST_LON = 90.0
FPS = 4
DPI = 130

# Same color scale/gradient as the Plate Carrée, Robinson, and Foucaut
# precipitation scripts, so this animation matches the others on the site.
# White at 0 cm blends into blue over just the first 0.3 cm, then continues
# through blue -> green -> yellow -> orange -> red -> purple up to 21 cm.
VMIN, VMAX = 0, 21
TICK_STEP = 3
color_stops = [
    (0 / VMAX, '#ffffff'),    # white           (0 cm)
    (0.3 / VMAX, '#d0e8f0'),  # pale blue       (0.3 cm)
    (2 / VMAX, '#7ab8e0'),    # light blue      (2 cm)
    (3 / VMAX, '#3060c0'),    # blue            (3 cm)
    (6 / VMAX, '#00994d'),    # green           (6 cm)
    (8 / VMAX, '#66cc00'),    # yellow-green    (8 cm)
    (9 / VMAX, '#ffff00'),    # yellow          (9 cm)
    (10 / VMAX, '#ff9900'),   # orange          (10 cm)
    (11 / VMAX, '#ff3300'),   # red-orange      (11 cm)
    (12 / VMAX, '#990000'),   # dark red        (12 cm)
    (21 / VMAX, '#660066'),   # purple          (21 cm)
]
CMAP = mcolors.LinearSegmentedColormap.from_list('custom_precip', color_stops, N=256)
BOUNDS = np.arange(VMIN, VMAX + 1, TICK_STEP)  # 0, 3, 6, ..., up to VMAX
if BOUNDS[-1] != VMAX:
    BOUNDS = np.append(BOUNDS, VMAX)  # always label the true top of the range


def find_precip_var(ds):
    candidates = [v for v in ds.data_vars if v != "frame_label"]
    if "tp_weekly_total_cm" in candidates:
        return "tp_weekly_total_cm"
    if len(candidates) == 1:
        return candidates[0]
    raise ValueError(f"Couldn't auto-pick the variable. Found: {candidates}.")


def build_hemisphere(lons, lats, central_lon):
    """Precompute everything that doesn't depend on frame data:
    the hemisphere mask, projected grid coords, boundary circle,
    and clipped/projected coastlines."""
    proj_str = f"+proj=nicol +lon_0={central_lon} +R=6371000"
    fwd = Transformer.from_crs("EPSG:4326", proj_str, always_xy=True)

    lon_diff = ((lons - central_lon + 180) % 360) - 180  # angular offset from center
    mask_1d = np.abs(lon_diff) <= 90
    if mask_1d.sum() < 2:
        raise RuntimeError(f"Hemisphere mask kept almost no columns for lon_0={central_lon}")

    # Sort the selected columns by their angular offset from the hemisphere
    # center, rather than trusting the original array's storage order. This
    # is what actually fixes the wraparound bug: lon_diff has no discontinuity
    # within a single +/-90deg window (its only jump is at +/-180 from center,
    # which is always excluded by the mask above), so this ordering is safe
    # for ANY central longitude and ANY input longitude convention -- unlike
    # a single global sort, which just relocates the seam onto whichever
    # hemisphere boundary happens to land on it.
    candidate_idx = np.where(mask_1d)[0]
    order = np.argsort(lon_diff[candidate_idx])
    col_index = candidate_idx[order]

    lons_hemi = lons[col_index]
    lon2d, lat2d = np.meshgrid(lons_hemi, lats)
    X, Y = fwd.transform(lon2d, lat2d)

    edge_lats = np.linspace(-90, 90, 400)
    bx1, by1 = fwd.transform(np.full_like(edge_lats, central_lon + 90), edge_lats)
    bx2, by2 = fwd.transform(np.full_like(edge_lats, central_lon - 90), edge_lats)
    boundary_x = np.concatenate([bx1, bx2[::-1]])
    boundary_y = np.concatenate([by1, by2[::-1]])

    shp_path = shpreader.natural_earth(resolution="110m", category="physical", name="coastline")
    reader = shpreader.Reader(shp_path)
    hemi_bbox = box(central_lon - 90, -90, central_lon + 90, 90)

    coastlines = []
    for record in reader.geometries():
        clipped = record.intersection(hemi_bbox)
        if clipped.is_empty:
            continue
        proj_geom = shp_transform(lambda x, y: fwd.transform(x, y), clipped)
        coastlines.append(proj_geom)

    # Country borders -- separate Natural Earth layer from coastlines
    # (admin_0_boundary_lines_land gives line geometries for country
    # borders directly, same structure as the coastline layer above, so
    # the same clip-then-reproject pipeline and plot_coastlines() helper
    # both work unchanged).
    shp_path_borders = shpreader.natural_earth(
        resolution="110m", category="cultural", name="admin_0_boundary_lines_land")
    reader_borders = shpreader.Reader(shp_path_borders)

    borders = []
    for record in reader_borders.geometries():
        clipped = record.intersection(hemi_bbox)
        if clipped.is_empty:
            continue
        proj_geom = shp_transform(lambda x, y: fwd.transform(x, y), clipped)
        borders.append(proj_geom)

    return {
        "col_index": col_index,
        "X": X,
        "Y": Y,
        "boundary_x": boundary_x,
        "boundary_y": boundary_y,
        "coastlines": coastlines,
        "borders": borders,
    }


def plot_coastlines(ax, coastlines, color="black", lw=0.5):
    for geom in coastlines:
        if geom.geom_type == "LineString":
            xs, ys = geom.xy
            ax.plot(xs, ys, color=color, linewidth=lw)
        elif geom.geom_type == "MultiLineString":
            for part in geom.geoms:
                xs, ys = part.xy
                ax.plot(xs, ys, color=color, linewidth=lw)


def main():
    print(f"Opening {NC_PATH}")
    ds = xr.open_dataset(NC_PATH)
    var_name = find_precip_var(ds)
    print(f"Using variable: '{var_name}'")

    da = ds[var_name]  # already in cm, no unit conversion needed
    lat_name = "latitude" if "latitude" in da.coords else "lat"
    lon_name = "longitude" if "longitude" in da.coords else "lon"

    lats = da[lat_name].values
    lons = da[lon_name].values
    n_frames = da.sizes["frame"]
    print(f"{n_frames} frames, grid {lats.size} x {lons.size}")

    frame_labels = ds['frame_label'].values if 'frame_label' in ds else None
    if frame_labels is None:
        print("WARNING: no 'frame_label' variable found -- falling back to frame index in titles.")

    # VMIN/VMAX are fixed (matching the other three precipitation scripts'
    # 0/21 scale), not derived from this data's actual range -- print a
    # heads-up if real values would exceed it, since pcolormesh silently
    # clips out-of-range values to the end color rather than erroring.
    actual_min, actual_max = float(da.min()), float(da.max())
    print(f"Actual data range: {actual_min:.2f} to {actual_max:.2f} cm "
          f"(color scale fixed at {VMIN} to {VMAX})")
    if actual_min < VMIN or actual_max > VMAX:
        print("  NOTE: some values fall outside the fixed color scale and will clip.")

    print("Precomputing Western Hemisphere geometry...")
    west = build_hemisphere(lons, lats, WEST_LON)
    print("Precomputing Eastern Hemisphere geometry...")
    east = build_hemisphere(lons, lats, EAST_LON)

    def render_frame(i):
        frame_data = da.isel(frame=i).values
        label = str(frame_labels[i]) if frame_labels is not None else f"Frame {i}"

        fig, axes = plt.subplots(1, 2, figsize=(13, 7), constrained_layout=True)

        for ax, hemi in [(axes[0], west), (axes[1], east)]:
            data_hemi = frame_data[:, hemi["col_index"]]
            mesh = ax.pcolormesh(hemi["X"], hemi["Y"], data_hemi,
                                  shading="auto", cmap=CMAP,
                                  vmin=VMIN, vmax=VMAX)
            plot_coastlines(ax, hemi["coastlines"], color="black", lw=1.0)
            plot_coastlines(ax, hemi["borders"], color="black", lw=0.8)
            ax.plot(hemi["boundary_x"], hemi["boundary_y"], color="black", linewidth=1)
            ax.set_aspect("equal")
            ax.axis("off")

        # Header text matching the other precipitation scripts' ERA-5/
        # Climate Reanalyzer style. Using ax.set_title(loc=...) instead of
        # fig.text() here -- constrained_layout only reserves vertical
        # space for real axes artists like titles, not for arbitrary
        # figure-coordinate text, so fig.text() was overlapping the top of
        # the circles.
        #
        # y is pinned explicitly (not left as the default "auto") because
        # auto-placement measures each axes' own content bounding box --
        # with axis("off") there's no spine to anchor to, and the two
        # hemisphere circles aren't pixel-identical in extent, so "auto"
        # put the two titles at very slightly different heights.
        TITLE_Y = 1.02
        TITLE_FONTSIZE = 16
        axes[0].set_title('ERA-5 | Climate Reanalyzer', loc='left', y=TITLE_Y,
                            fontsize=TITLE_FONTSIZE, fontweight='bold')
        axes[1].set_title(f'{label}; 1979\u20132000 Mean Weekly Total', loc='right', y=TITLE_Y,
                            fontsize=TITLE_FONTSIZE, fontweight='bold')

        cbar = fig.colorbar(mesh, ax=axes, orientation='horizontal',
                             pad=0.08, fraction=0.12, shrink=0.5,
                             aspect=18, extend='neither')
        cbar.set_label('Precipitation (cm)', fontsize=14)
        cbar.ax.tick_params(labelsize=12)
        cbar.set_ticks(BOUNDS)
        cbar.set_ticklabels([str(t) for t in BOUNDS])

        buf = io.BytesIO()
        # NOTE: no bbox_inches="tight" here -- it makes each frame a
        # different pixel size depending on content extents, which
        # breaks the video encoder (needs identical, even dimensions
        # every frame).
        fig.savefig(buf, format="png", dpi=DPI)
        plt.close(fig)
        buf.seek(0)
        img = imageio.imread(buf)

        # drop alpha channel (write_frames wants RGB24) and force even
        # width/height (required by libx264)
        img = img[:, :, :3]
        h, w = img.shape[:2]
        img = img[: h - (h % 2), : w - (w % 2)]
        return np.ascontiguousarray(img)

    # Render frame 0 first to lock in the output size, then start the
    # ffmpeg process directly via imageio_ffmpeg -- this bypasses
    # imageio.get_writer()'s plugin auto-resolution, which on some
    # installs picks the tifffile backend for .mp4 output instead of
    # ffmpeg (that backend doesn't accept fps=, hence the TypeError).
    print("Rendering frame 0...")
    first_frame = render_frame(0)
    height, width = first_frame.shape[:2]

    ffmpeg_writer = imageio_ffmpeg.write_frames(
        OUT_MP4_PATH, size=(width, height), fps=FPS,
        codec="libx264", quality=8, macro_block_size=1,
    )
    ffmpeg_writer.send(None)  # seed/initialize the generator
    ffmpeg_writer.send(first_frame.tobytes())
    print(f"  frame 1/{n_frames} done")

    for i in range(1, n_frames):
        frame_img = render_frame(i)
        ffmpeg_writer.send(frame_img.tobytes())
        if i % 8 == 0:
            print(f"  frame {i + 1}/{n_frames} done")

    ffmpeg_writer.close()
    print(f"Saved: {OUT_MP4_PATH}")


if __name__ == "__main__":
    main()